### ISSUE TITLE
code that is not tested should not be included in reports

### ISSUE THREAD

**Issue by xh3b4sd** [codecov/support#154]
I recently noticed that some reports are broken and even flap in how they report. See https://codecov.io/github/xh3b4sd/anna/common.go?ref=675f25074c790af37089716b4af701bf3b82eef2. There is no test testing this code. Usually code that is not tested does not show up in the report. Further the highlighting/calculation is confusing, because a few lines of codes are considered, and others are not. This looks broken to me. Any information on this?

**Comment by xh3b4sd** [comment]
Now the report states the files coverage is 100%. See https://codecov.io/github/xh3b4sd/anna/common.go?ref=c639cfd64f7c6bab3eea174327e826f9e8fd01e3.

This is simply wrong, because there is still no test file for it. The fact the report is flapping so much makes the coverage service pretty confusing and even useless. At the moment I don't know what is right or wrong. 

It would be very much appreciated if somebody can help here.

**Comment by stevepeak** [comment]
Hey @xh3b4sd this issue is language specific. Codecov process the reports given and some languages do not include files that were not imported. 

Unfortunately there is no way to provide a fix for this because Codecov is not able to recognize the files without a provided coverage report. Simply put: we express exactly what is reported. 

I would suggest taking this issue up with the Go-cover team to see why the collector is ignoring these lines. Thank you

**Comment by xh3b4sd** [comment]
I don't think it is a language specific error. When codecov reports what it receives there should not be any coverage report for the file we are talking about here. Lets put it together again:

- There is no test that tests code within github/xh3b4sd/anna/common.go.
- Codecov shows that github/xh3b4sd/anna/common.go is tested and covered though. This is simply wrong.
- I have attached a [coverage.txt](https://github.com/codecov/support/files/224463/coverage.txt) file from the current master. In there is not a single line mentioning statements of the file github/xh3b4sd/anna/common.go. Thus the go tool seems to be correct.

Another thing. You mentioned go cover. I don't use this. I use the following script that simply uses go test. This is basically what the codecov guides suggest but only using a modified find so only packages are run that actually contain tests. I am using this from the very beginning and it worked all the time, except since a few days/weeks.

```
$ cat go.test.sh
#!/usr/bin/env bash

set -e
echo "" > coverage.txt

for d in $(find ./* -maxdepth 10 -type f -name '*test.go' -not -path "./.workspace/*" -not -path "./.git/*" -exec dirname {} \; | sort -u); do
  go test -coverprofile=profile.out -covermode=atomic $d
  if [ -f profile.out ]; then
      cat profile.out >> coverage.txt
      rm profile.out
  fi
done

**Comment by xh3b4sd** [comment]
Another weird thing is that when running tests in master locally and looking at what `go test -cover` outputs, the result differs from codecov's interpretation partially quite a lot. 

```
$ make gotest
ok  	_/home/vagrant/projects/private/anna/clg	0.976s	coverage: 99.7% of statements
ok  	_/home/vagrant/projects/private/anna/client/control/log	0.028s	coverage: 96.0% of statements
ok  	_/home/vagrant/projects/private/anna/client/interface/text	0.017s	coverage: 98.1% of statements
ok  	_/home/vagrant/projects/private/anna/distribution	0.007s	coverage: 79.6% of statements
ok  	_/home/vagrant/projects/private/anna/factory/client	0.007s	coverage: 100.0% of statements
ok  	_/home/vagrant/projects/private/anna/factory/server	0.010s	coverage: 96.1% of statements
ok  	_/home/vagrant/projects/private/anna/feature-set	0.139s	coverage: 64.3% of statements
ok  	_/home/vagrant/projects/private/anna/gateway	0.006s	coverage: 100.0% of statements
ok  	_/home/vagrant/projects/private/anna/id	0.040s	coverage: 87.5% of statements
ok  	_/home/vagrant/projects/private/anna/log	0.007s	coverage: 99.0% of statements
ok  	_/home/vagrant/projects/private/anna/net/pat	0.006s	coverage: 90.0% of statements
ok  	_/home/vagrant/projects/private/anna/scheduler	0.362s	coverage: 92.1% of statements
ok  	_/home/vagrant/projects/private/anna/storage/memory	0.008s	coverage: 98.0% of statements
ok  	_/home/vagrant/projects/private/anna/storage/redis	0.015s	coverage: 96.8% of statements
ok  	_/home/vagrant/projects/private/anna/strategy	0.007s	coverage: 98.5% of statements
```

Note that the `distribution` package here states nearly 80%. Looking at codecov's reports in master it states nearly 100%. You can do the math. See https://codecov.io/github/xh3b4sd/anna/distribution?ref=c639cfd64f7c6bab3eea174327e826f9e8fd01e3

So there is definitely something fishy. Interestingly that the differences (always?) have to do with the same file name. That is `common.go`. In the projects root this file is added to the report but should not be included. In the distribution package there is a `common.go` file ignored. Very strange. 

Hope this helps.

**Comment by xh3b4sd** [comment]
Can anybody help here?

**Comment by xh3b4sd** [comment]
FYI I just moved code around and the coverage reported by codecov for the distribution packages was set down to about 93%. Interestingly now the `common.go` file within this specific package is now included in the report. Before it was missing. After the code move it is in. This is definitely not golang specific since the coverage reported by `go test` stayed the same.

**Comment by stevepeak** [comment]
Hey @xh3b4sd I'll jump in here to help you out very soon, I'm a little behind on support because of a scheduled release tomorrow that has taken a lot of my attention. Thank you for your patience 👍

**Comment by xh3b4sd** [comment]
Thanks for your reply. Don't worry. All good. As long as this is covered somehow at some point of time, I am fine.

**Comment by xh3b4sd** [comment]
This issue is still relevant. Is there any update you can provide?

**Comment by stevepeak** [comment]
Hey @xh3b4sd Can I review a more recent commit? Let's use https://codecov.io/github/xh3b4sd/anna/tree/5fd63b3

Can you point out some issues with this build. Thanks!

**Comment by xh3b4sd** [comment]
Take a look at `common.go`. 

At first the highlighting of the code makes no sense to me. It implies that a few lines of a method are covered and the very next lines are not. This is impossible. There are no conditional branches. E.g. line 35 is covered. The next effective code line is 37. This is not covered though. 

Second, there is no single test that tests the code of `common.go`. So it is simply wrong that there is any coverage report for this file. Evidence for this is also the `coverage.txt` containing coverage statements of the whole test suite. There is no single mention for this file. 

The same issues apply to `error.go` and `object.go`.

The issue also takes place the other way around. E.g. `factory/id/error.go` is tested, but no coverage is shown for that. The only coverage reported for the `factory/id` package is for `factory/id/factory.go`. Here the `coverage.txt` contains statements for `factory/id/error.go`.

**Comment by stevepeak** [comment]
Ok. Found an issue with mapping paths back to the proper relative files. Fixing now

**Comment by stevepeak** [comment]
Deployed! Can you push again?

**Comment by xh3b4sd** [comment]
Looking at the new commit the misleading file reports seem to be fixed. https://codecov.io/gh/xh3b4sd/anna/tree/4bdee768f37f778c7e0f787ad3fd4595571c77e5

I don't get hook updates in my PR and every now and then I get 404s in the codecov webinterface, but these are separate issues.
